[WIP] JIT Static Hooks: schema checking logic

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49547 [WIP] JIT Static Hooks: cpp tests
* #49546 [WIP] JIT Static Hooks: python tests
* #49545 [WIP] JIT Static Hooks: serialization logic
* **#49975 [WIP] JIT Static Hooks: schema checking logic**
* #49544 [WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation

Added error messages that get printed if a hook has an incorrect schema. 
The error messages print out specifically what is wrong with the schema with a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

Commit:
[WIP] JIT Static Hooks: schema checking logic

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed during the compilation of hooks. Modified error_report.h/.cpp to print out 'hints' to give hints of what could be wrong. For the case of hooks, if an error is thrown during the scripting of a hook, the error message will contain a hint with information telling the user that the issue arose when a hook was being compiled. 

The error messages for schema issues print out specifically what is wrong with the schema and then a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed during the compilation of hooks. Modified error_report.h/.cpp to print out 'hints' to give hints of what could be wrong. For the case of hooks, if an error is thrown during the scripting of a hook, the error message will contain a hint with information telling the user that the issue arose when a hook was being compiled. 

The error messages for schema issues print out specifically what is wrong with the schema and then a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed during the compilation of hooks. Modified error_report.h/.cpp to print out 'hints' to give hints of what could be wrong. For the case of hooks, if an error is thrown during the scripting of a hook, the error message will contain a hint with information telling the user that the issue arose when a hook was being compiled. 

The error messages for schema issues print out specifically what is wrong with the schema and then a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed during the compilation of hooks. Modified error_report.h/.cpp to print out 'hints' to give hints of what could be wrong. For the case of hooks, if an error is thrown during the scripting of a hook, the error message will contain a hint with information telling the user that the issue arose when a hook was being compiled. 

The error messages for schema issues print out specifically what is wrong with the schema and then a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed during the compilation of hooks. Modified error_report.h/.cpp to print out 'hints' to give hints of what could be wrong. For the case of hooks, if an error is thrown during the scripting of a hook, the error message will contain a hint with information telling the user that the issue arose when a hook was being compiled. 

The error messages for schema issues print out specifically what is wrong with the schema and then a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed during the compilation of hooks. Modified error_report.h/.cpp to print out 'hints' to give hints of what could be wrong. For the case of hooks, if an error is thrown during the scripting of a hook, the error message will contain a hint with information telling the user that the issue arose when a hook was being compiled. 

The error messages for schema issues print out specifically what is wrong with the schema and then a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed during the compilation of hooks. Modified error_report.h/.cpp to print out 'hints' to give hints of what could be wrong. For the case of hooks, if an error is thrown during the scripting of a hook, the error message will contain a hint with information telling the user that the issue arose when a hook was being compiled. 

The error messages for schema issues print out specifically what is wrong with the schema and then a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

[ghstack-poisoned]

SplitInfinity reviewed (commented):
Could you elaborate a bit on the purpose and intended use of `HintStack` (especially how it can be used in general, outside of the context of this feature)? It seems to be used as a place to store some extra error messages about schema mismatches so that they can be used when the schema is compiled if it turns out the hook schema is not what we expect.

I think we can make do without it by calling `ensure_defined()` on compiled hook functions in `define_hooks` and then checking the schemas (which can be accessed through `function->getSchema()`). That way we know for a fact that the function has been compiled, and don't need to save the error messages away somewhere to be used later. In addition, we can use the `SourceRange` on the `Def` for the hook to highlight the signature of the hook function in the error message. 

What do you think? At this point, you're more aware of the nitty gritty details than I am, and I bet you tried this.

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed during the compilation of hooks. Modified error_report.h/.cpp to print out 'hints' to give hints of what could be wrong. For the case of hooks, if an error is thrown during the scripting of a hook, the error message will contain a hint with information telling the user that the issue arose when a hook was being compiled. 

The error messages for schema issues print out specifically what is wrong with the schema and then a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

[ghstack-poisoned]

Lilyjjo commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1079,6 +1079,299 @@ torch::jit::Function* ClassType::findForwardHook(const std::string& name) const
+  for (int i = 1; i < forward_args.size(); ++i) {
+    input_types << forward_args[i].type()->annotation_str();
+    if (forward_args.size() - 1 != i) {
+      input_types << " ";
```
It would've been interesting if you did because I forgot to add them 😬

Lilyjjo commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1079,6 +1079,299 @@ torch::jit::Function* ClassType::findForwardHook(const std::string& name) const
+  const Argument& pre_output =
+      (forward_hooks_.size() == 1)
+          ? forwardSchema.returns()[0]
+          : forward_hooks_[forward_hooks_.size() - 2]->getSchema().returns()[0];
```
You're right. I changed it to take the hook's index. It needs to be more specific than just the hook with the same name because the same hook can be called multiple time.

Lilyjjo commented:
> Could you elaborate a bit on the purpose and intended use of `HintStack` (especially how it can be used in general, outside of the context of this feature)? It seems to be used as a place to store some extra error messages about schema mismatches so that they can be used when the schema is compiled if it turns out the hook schema is not what we expect.
I agree this solution isn't clean. The issue I was trying to get around is more driven by two usability concerns than actual need. I was concerned that users would be confused by the new behavior of hooks being scripted when they weren't before in general, this error message would inform them if that they were scripting hooks in the first place. If an attached hook has `ensure_defined()` called on it and errors out it'll just print out a normal error message with no mention of hooks without the `HintStack` functionality. More concerning is that the schema itself won't be checked until the hook is totally scriptable, so users could spend time fixing the hooks to be scriptable to match an incorrect schema which they could find annoying (I'd find it annoying). 

I don't know of any specific places where the `HintStack` can be used, I was going off of Zach's comment in the original PR saying 'It would be better to augment the call stack with a generic facility put more information into the stack trace when an error occurs'. Maybe it'd be helpful for future error messages? I tried out several different versions of modifying the call stack and this was the cleanest solution I could come up with that avoided major refactoring. I can look into doing a major refactoring to make it cleaner if you think that it'd be worth it to have the hook information message printed for the users.

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed during the compilation of hooks. Modified error_report.h/.cpp to print out 'hints' to give hints of what could be wrong. For the case of hooks, if an error is thrown during the scripting of a hook, the error message will contain a hint with information telling the user that the issue arose when a hook was being compiled. 

The error messages for schema issues print out specifically what is wrong with the schema and then a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

[ghstack-poisoned]

SplitInfinity commented:
> I agree this solution isn't clean. The issue I was trying to get around is more driven by two usability concerns than actual need. In general I was concerned that users would be confused by the new behavior of hooks being scripted when they weren't before. In `ensure_defined()` the hook's schema isn't accessible until the hook itself is scriptable. If the hook errors out during scripting it'll just print out a normal error message with no mention of hooks without the `HintStack` functionality that'll inform the users that they are scripting hooks. The second concerning thing is that the schema itself won't be checked until the hook is totally scriptable, so users could spend time fixing the hooks to be scriptable to match an incorrect schema which they could find annoying (I'd find it annoying).

Very good points. Another option that comes to my mind is creating a local `ScriptTypeParser` instance within the scope of `define_hooks` and using that to parse the schema for the hook (`parseSchemaFromDef`) before attempting to script it. It is duplicating work a little, but at least that way the schema can be accessed without trying to compile the entire function. And if the schema cannot be parsed, you know that you're in `define_hooks` so you can throw a nice error. This still leaves unresolved the issue of informing the user that the hooks of their module are also being scripted. To be honest, I think we should get more opinions from others on `HintStack` and leave this unresolved for now.

Lilyjjo commented:
> Another option that comes to my mind is creating a local ScriptTypeParser instance within the scope of define_hooks and using that to parse the schema for the hook (parseSchemaFromDef) before attempting to script it.
This option doesn't work because non-obviously `parseSchemaFromDef` won't capture non-annotated return types. The `emitDef(...)` that is used in ir_emitter.cpp to create the schema sets the returns in the schema based off of the emitted outputs. We could solve this issue by requiring hook signatures to be annotated with a return type on the python side

SplitInfinity commented:
> > Another option that comes to my mind is creating a local ScriptTypeParser instance within the scope of define_hooks and using that to parse the schema for the hook (parseSchemaFromDef) before attempting to script it.
> 
> This option doesn't work with the current required hook signatures because `parseSchemaFromDef` won't capture non-annotated return types. The `emitDef(...)` that is used in ir_emitter.cpp to create the schema sets the returns in the schema based off of the emitted outputs. We could solve this issue by requiring hook signatures to be annotated with a return type on the python side.
> 
> Most likely requiring users to annotate the return type would probably catch most unintended hook additions. I doubt that many users who are adding hooks to eager mode would randomly also include return type annotations.

I think we should go with this for now and remove the `HintStack` and think about that some more. I think we can work on it under the general task of inferring pre-hook input and output types and hook input types from `forward`.

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed during the compilation of hooks. Modified error_report.h/.cpp to print out 'hints' to give hints of what could be wrong. For the case of hooks, if an error is thrown during the scripting of a hook, the error message will contain a hint with information telling the user that the issue arose when a hook was being compiled. 

The error messages for schema issues print out specifically what is wrong with the schema and then a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed during the compilation of hooks. Modified error_report.h/.cpp to print out 'hints' to give hints of what could be wrong. For the case of hooks, if an error is thrown during the scripting of a hook, the error message will contain a hint with information telling the user that the issue arose when a hook was being compiled. 

The error messages for schema issues print out specifically what is wrong with the schema and then a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed during the compilation of hooks. Modified error_report.h/.cpp to print out 'hints' to give hints of what could be wrong. For the case of hooks, if an error is thrown during the scripting of a hook, the error message will contain a hint with information telling the user that the issue arose when a hook was being compiled. 

The error messages for schema issues print out specifically what is wrong with the schema and then a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed if a hook has an incorrect schema. 
The error messages print out specifically what is wrong with the schema with a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed if a hook has an incorrect schema. 
The error messages print out specifically what is wrong with the schema with a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed if a hook has an incorrect schema. 
The error messages print out specifically what is wrong with the schema with a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed if a hook has an incorrect schema. 
The error messages print out specifically what is wrong with the schema with a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed if a hook has an incorrect schema. 
The error messages print out specifically what is wrong with the schema with a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed if a hook has an incorrect schema. 
The error messages print out specifically what is wrong with the schema with a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

[ghstack-poisoned]

SplitInfinity reviewed (approved):
We should try to figure out a way to use the schema from `forward` to infer the input type and output types of all pre-hooks and hooks. 

What I had in mind is the following. We can propagate the input type of `forward` to the `input` parameter of all pre-hooks and hooks (after wrapping it in a `Tuple` as needed). We can infer the type of the `output` parameter of the first forward hook based on the output type of `forward`. Based on the inferred output type of forward hook `n`, we should be able to infer the type of the `output` argument of forward hook `n+1`. And we should also figure out how to extend `ErrorReport` to provide a good user experience when these type inference shenanigans cause hooks to fail to compile.

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed if a hook has an incorrect schema. 
The error messages print out specifically what is wrong with the schema with a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed if a hook has an incorrect schema. 
The error messages print out specifically what is wrong with the schema with a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: schema checking logic"


Added error messages that get printed if a hook has an incorrect schema. 
The error messages print out specifically what is wrong with the schema with a more generic message with an example of the correct schema. Example of wrong output type on a forward hook: 

"Hook 'forward_hook' on module 'OuterModuleSingleIO' has the wrong type for the output argument. Recieved type: 'str'. Expected type: 'Tuple[str]'.
This error occured while scripting the forward hook 'forward_hook' on module OuterModuleSingleIO. If you did not want to script this hook remove it from the original NN module before scripting. This hook was expected to have the following signature: forward_hook(self, input: Tuple[str], output: Tuple[str]). The type of the output arg is the returned type from either the forward method or the previous hook if it exists. Note that hooks can return anything, but if the hook is on a submodule the outer module is expecting the same return type as the submodule's forward."

Differential Revision: [D25771120](https://our.internmc.facebook.com/intern/diff/D25771120)

[ghstack-poisoned]
